### ISSUE TITLE
decimal can only support precision up to 38

### ISSUE THREAD

**Issue by KalaniStanton** [moj-analytical-services/splink#694]
### What happens?

Error is that the SQL backend for pyspark only supports up to 38 decimals. This error is referenced in issue #48  and was marked as resolved, but I am still getting this issue when trying to run the second `estimate_parameters_using_expectation_maximisation` function below.

```python
blocking_rule_for_training_fname = "l.Fname = r.Fname and l.Lname = r.Lname"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training_fname)

blocking_rule_for_training_lname = "l.State = r.State"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training_lname)
```

### To Reproduce

Opened up small dataproc cluster on GCP for pyspark with the following console commands in region useast-1 & useast-4
```bash
gcloud beta dataproc clusters create ${CLUSTER_NAME} --bucket bucket_for_spark_test --worker-machine-type n1-standard-4 --master-boot-disk-size 500 --num-workers 4 --worker-boot-disk-type pd-ssd --worker-boot-disk-size 100 --num-worker-local-ssds 2 --image-version 1.5-debian --initialization-actions gs://dataproc-initialization-actions/python/pip-install.sh --metadata 'PIP_PACKAGES=google-cloud-storage splink==3.1.0 pandas scipy' --max-idle 7200s --scopes 'https://www.googleapis.com/auth/cloud-platform' --enable-component-gateway --project ${PROJECT_ID}
```

I then run the script stored in a test_splink_job.py script by running the following:
```bash
cd ~/pyspark-tests 
gcloud dataproc jobs submit pyspark --cluster ${CLUSTER_NAME} \
		--jars gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar \
	    --driver-log-levels root=FATAL \
	    test_splink_job.py
```

### Datasets 

The data being used in the spark job is two datasets with identically named columns from two data sources each containing 50 records.

### Python Code
The *python code* in that script prior to the error:
```python
from pyspark.sql.types import StructField, StructType, StringType, LongType
from pyspark.sql import SparkSession
from py4j.protocol import Py4JJavaError

import os
spark = SparkSession.builder.appName("record-linkage").config("spark.sql.crossJoin.enabled", "true").getOrCreate()

# Import Splink
from splink.spark.spark_linker import SparkLinker
import splink.spark.spark_comparison_library as cl
import splink.spark.spark_comparison_level_library as cll
from splink.charts import save_offline_chart

bucket_for_files = 'gs://bucket_for_spark_test'

voter_df = (spark.read.format('bigquery').option('table', f"{project_id}.test_set.voters")
                    .load())
commercial_df = (spark.read.format('bigquery').option('table', f"{project_id}.test_set.commercials")
                    .load())

settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [
        "l.Fname = r.Fname"
    ],
    "comparisons": [
        cl.levenshtein_at_thresholds("Fname", 2),
        cl.exact_match("Lname", term_frequency_adjustments=True),
        cl.exact_match("birthday"),
        cl.exact_match("Primary_Address"),
        cl.exact_match("City_Name"),
        cl.exact_match("State"),
        cl.exact_match("Cell_Phone")
    ],
}

linker = SparkLinker([voter_df, commercial_df], settings, input_table_aliases=table_names)

linker.estimate_u_using_random_sampling(target_rows=25000)

blocking_rule_for_training_fname = "l.Fname = r.Fname and l.Lname = r.Lname"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training_fname)

blocking_rule_for_training_lname = "l.State = r.State"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training_lname) # This is where error occurs
```

### Error 
```bash
Traceback (most recent call last):
  File "/tmp/4885fdbf55f9475c9669e790e97b3bab/test_spark_job.py", line 64, in <module>
    linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training_fname)
  File "/opt/conda/default/lib/python3.7/site-packages/splink/linker.py", line 915, in estimate_parameters_using_expectation_maximisation
    em_training_session._train()
  File "/opt/conda/default/lib/python3.7/site-packages/splink/em_training_session.py", line 169, in _train
    expectation_maximisation(self, cvv)
  File "/opt/conda/default/lib/python3.7/site-packages/splink/expectation_maximisation.py", line 167, in expectation_maximisation
    df_params = linker._execute_sql_pipeline([df_comparison_vector_values])
  File "/opt/conda/default/lib/python3.7/site-packages/splink/linker.py", line 328, in _execute_sql_pipeline
    transpile,
  File "/opt/conda/default/lib/python3.7/site-packages/splink/linker.py", line 406, in _sql_to_splink_dataframe
    sql, output_tablename_templated, table_name_hash, transpile=transpile
  File "/opt/conda/default/lib/python3.7/site-packages/splink/spark/spark_linker.py", line 206, in _execute_sql
    spark_df = self.spark.sql(sql)
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/session.py", line 767, in sql
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1257, in __call__
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/utils.py", line 73, in deco
pyspark.sql.utils.ParseException: "\ndecimal can only support precision up to 38
```

I have omitted the query from the text, but I see multiple values within the query that are of precision >38. Would be happy to provide that full query if requested.

### OS:

Linux - Dataproc on GCP (see To Reproduce text for config)

### Splink version:

3.1.0

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by RobinL** [comment]
Thanks for the detailed info.  We'll look into this.  Just adding some notes from some initial reserach:

In Splink 2, we seemed to avoid this issue using [this syntax](https://github.com/moj-analytical-services/splink/commit/7b9f76f05bbab257635e539e623656db61a0d7fc#diff-b34c145bfb205605355ac0e86b057c4a7fbde75238b9e2961a71f20ba4b0c345R226
).

i.e. either of the following two worked:
```
{prob}D
```
or
```
cast({prob:.35f} as double)
```


I’m not sure why using the `D` syntax meant we could get rid of the `.35f` syntax.

More info on the `D` syntax [here](https://github.com/tobymao/sqlglot/issues/58) but not sure it offers any clues why it solved the `decimal can only support precision up to 38` problem

A bit unclear what's the cleanest solution here.  would be nice to avoid having to use `.35f` if possible, but ultimately it shouldn't affect accuracy in any meaningful way.

**Comment by ThomasHepworth** [comment]
`sqlglot` automatically transpiles `0.1234D` -> `cast(0.1234 as double)`, so we'd need to tinker with [this](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/spark.py) to get it working. 

Probably easiest to just add a quick translation step for spark that caps our decimal places whenever we see a double.

<hr>

**sqlglot transpiling example:**

**Input**
```
import sqlglot
sql = "select 0.0055691509084167595D"
sqlglot.parse_one(sql, read="spark").sql()
```

**Out**

`'SELECT CAST(0.0055691509084167595 AS DOUBLE)'`

**Comment by ThomasHepworth** [comment]
Fixed in [PR #697](https://github.com/moj-analytical-services/splink/pull/697)

**Comment by RobinL** [comment]
Further info

```spark.sql('select 0.1 as a').toPandas().iloc[0,0]`
-> Decimal('0.1')```

```spark.sql('select 0.12345678D as a').toPandas().iloc[0,0]
->  0.12345678. (i.e. a float)```

```spark.sql('select cast(0.349807234523750923475089457470478952340895740 as double) as a').toPandas().iloc[0,0]
-> ArithmeticException: [DECIMAL_PRECISION_EXCEEDS_MAX_PRECISION] Decimal precision 45 exceeds max precision 38.```

```spark.sql('select 0.349807234523750923475089457470478952340895740D as a').toPandas().iloc[0,0]

0.3498072345237509```

Conclusion: we need to use `{num}D` for inserting a literal a bug free !

At the moment, that’s what the [custom spark dialect](https://github.com/moj-analytical-services/splink/blob/663491b91bc7d0888ba27e76ebd0466b1bf3615d/splink/spark/spark_helpers/custom_spark_dialect.py#L22-L23) does - i.e. it [transpiles](https://github.com/moj-analytical-services/splink/blob/033b04168103deca6bcc7522cbdbe6e6db84fe68/splink/spark/linker.py#L399)  things like [THEN cast({bayes_factor} as float8)](https://github.com/moj-analytical-services/splink/blob/663491b91bc7d0888ba27e76ebd0466b1bf3615d/splink/comparison_level.py#L557) into `{bayes_factor}D`.

I think in Splink 4 we probably just want the syntax to be part of the dialects, and to do this explicitly rather than use the custom transpilation
